lib/Sema/SemaBounds.cpp

```diff
@@ -578,6 +578,243 @@ namespace {
+      if (!NeedleDeclRef) {
+        // Not a DeclRef. Error, stop
+        S.Diag(Needle->getExprLoc(), diag::err_cast_to_checked_fn_ptr_must_be_named)
+          << ToType << E->getSourceRange();
+
+        return;
+      }
+
+      const FunctionDecl *NeedleFun = dyn_cast<FunctionDecl>(NeedleDeclRef->getDecl());
+      if (!NeedleFun) {
+        // Not a DeclRef to a Top-Level function. Error, stop.
+        S.Diag(Needle->getExprLoc(), diag::err_cast_to_checked_fn_ptr_must_be_named)
+          << ToType << E->getSourceRange();
+
+        return;
+      }
+      QualType NeedleFunType = NeedleFun->getType();
+
+      if (!S.Context.typesAreCompatible(
+        ToType->getPointeeType(), 
+        NeedleFunType,
+        /*CompareUnqualified=*/false,
+        /*IgnoreBounds=*/false)) {
+        // The type of the defined function is not compatible with the pointer
+        // we're trying to assign it to. Error, stop.
+
+        S.Diag(Needle->getExprLoc(), diag::err_cast_to_checked_fn_ptr_from_incompatible_type)
+          << ToType << NeedleFunType << false << E->getSourceRange();
+
+        return;
+      }
+
+    }
+
+    // This is used in void CheckDisallowedFunctionPtrCasts(Expr*)
+    // to find if a cast is value-preserving
+    //
+    // Other operations might also be, but this algorithm is currently
+    // conservative.
+    //
+    // This will add the required error messages
+    bool CheckValuePreservingCast(const CastExpr *E, const QualType ToType) {
+      switch (E->getCastKind())
+      {
+      case CK_NoOp:
+      case CK_NullToPointer:
+      case CK_FunctionToPointerDecay:
+      case CK_BitCast:
+        return true;
+      case CK_LValueToRValue: {
+        // Reads of checked function pointers are allowed
+        QualType ETy = E->getType();
+        if (ETy->isCheckedPointerPtrType() &&
+          ETy->isFunctionPointerType())
+          return true;
+
+        // This reads unchecked memory, which is definitely not value-preserving
+        S.Diag(E->getExprLoc(), diag::err_cast_to_checked_fn_ptr_cannot_read_mem)
+          << ToType << E->getSourceRange();
+
+        return false;
+      }
+      default:
+        S.Diag(E->getExprLoc(), diag::err_cast_to_checked_fn_ptr_not_value_preserving)
+          << ToType << E->getSourceRange();
+
+        return false;
+      }
+    }
+
+    // This is used in void CheckDisallowedFunctionPtrCasts(Expr*)
+    // to find if the thing we just discovered is deref (*) or
+    // addr-of (&) operator on a function pointer type.
+    // These operations are value perserving.
+    //
+    // Other operations might also be, but this algorithm is currently
+    // conservative.
+    //
+    // This will add the required error messages
+    bool CheckValuePreservingCastLikeOp(const UnaryOperator *E, const QualType ToType) {
+      QualType ETy = E->getType();
+      QualType SETy = E->getSubExpr()->getType();
+
+      bool between_functions = false;
```

Comment:
You can just have two local variables that are scope by the bodies of the case statement.